src/node/lattice.rs

```diff
@@ -0,0 +1,518 @@
+        let mut dfs = DfsPostOrder::empty(&*forest);
+        let mut dependencies: HashSet<NodeIndex<u32>> = HashSet::new();
+        let mut add_edges_to: Vec<NodeIndex<u32>> = Vec::new();
+        let mut add_edges_from: Vec<NodeIndex<u32>> = Vec::new();
+        let mut remove_roots: Vec<(usize, NodeIndex<u32>)> = Vec::new();
+
+        for root in &*roots {
+            // This function maintains the stack when iterating over the graph so we do not have to
+            // check the same node in the graph twice.
+            dfs.move_to(*root);
+
+            // Retrieve the next node in DFS order from the forest.
+            while let Some(nx) = dfs.next(&*forest) {
+                // Check if any of the children of the current node in DFS traversal already have a
+                // dependency on the node we are trying to add.
+                let mut dependent_child: bool = false;
+                for child in forest.neighbors(nx) {
+                    if dependencies.contains(&child) {
+                        dependencies.insert(nx);
+                        dependent_child = true;
+                        break;
+                    }
+                }
+
+                // If the event we are trying to add has already been added as a dependency of one
+                // of the children of the DFS node, then move to the next root of the forest.
+                if dependent_child {
+                    break;
+                }
+
+                if let Some(node) = forest.node_weight(nx).unwrap().as_ref() {
+                    // If the event we are trying to add is dependent on the DFS node, then
+                    // add an edge from the DFS node to the event to be added.
+                    if event < *node || event == *node {
+                        // Check if any of the children of the DFS node should be dependent on the
+                        // event we are trying to add.
+                        let mut remove_edges: Vec<NodeIndex<u32>> = Vec::new();
+                        for child in forest.neighbors(nx) {
+                            let child_node: &OperatorEvent =
+                                &forest.node_weight(child).unwrap().as_ref().unwrap();
+                            if *child_node < event {
+                                // The child has a dependency on the node being added. 
+                                // Break the dependency from the DFS node to its child, and add a
+                                // dependency from the node to be added to the child.
+                                add_edges_to.push(child);
+                                remove_edges.push(child);
+                            }
+                        }
+
+                        // Add a dependency from the DFS node to the node being added, only if the
+                        // partial order says so.
+                        if event < *node {
```

Comment:
If you don't want to deref
```suggestion
                        if &event < node {
```